Cookbook/Proposals/FeatureSwitchesRedesign.md

```diff
@@ -0,0 +1,218 @@
+    let name: String
+    // key to query the value from provider and key for UserDefaults
+    let key: String
+}
+
+public struct LocalFeatureSwitch<T> {
+    let provider = Current.userDefaults
+    var value: T { ... }
+    let name: String
+    let key: String
+}
+
+public struct ProductFeatureSwitch<T> {
+    let provider = Current.productConfigurator
+    var value: T { ... }
+    let name: String
+    let key: String
+}
+```
+
+
+Using dedicated switch types we will be also able to explicitly specify what remote feature switches are A/B tested and which not:
+
+```swift
+typealias ABTest = RemoteFeatureSwitch
+```
+
+Remote Feature Switches we fetch from Firebase all together and under the hood they are just abstract key-value pairs. So to access them in a module the module should be initialised with this abstract collection of key-value pairs (or access it via `Current` as in examples) so that it can get from it values it cares about and store them for later use.
+
+These feature switches should be separate from static configuration (implemented ATM via `AppConfiguration`) but both should be defined and stored inside the module. This way we break up the monolythic `AppConfiguration` and allow static configuration be configured separately from dynamic feature switches which get their values at runtime.
+
+With patient details situation is a bit different because this data is associated with a `PatientDTO`. We can refactor it so that `PatientDTO` stores abstract key-value pairs and pass it to the frameworks after patient details are retrieved similarly with Firebase remote config (or similarly access it via `Current` when these patient details will be moved there). But we suggest to address this type of feature switches separately as this work depends on other work related to how the app works with patient details. In future these feature switches can become just a new type of switch like `PatientFeatureSwitch` which will use `Current.patient.details` as their provider.
+
+Inside the module we can extend `World` with a module specific namespace and keep its feature switches and configurations separate from the rest of `World` properties.
+
+```swift
+// In the feature framework
+extension Current {
+    private static private(set) var someFeature: SomeFeatureModule!
+    
+    // way for the app to set static configurations for the feature
+    public func configureSomeFeature(
+        config: SomeFeatureModule.Configuration
+    ) {
+        World.someFeature = SomeFeatureModel(config: config)
+    }
+    
+    public var someFeature: FeatureModule {
+        return World.someFeature
+    }
+    
+    public struct SomeFeatureModule {
+       
+        public let featureSwitches: FeatureSwitches
+    
+        // dynamic feature switches
+        public struct FeatureSwitches {
+            public let isNewFeatureEnabled = RemoteFeatureSwitch<Bool>(
+                name: “New feature”,
+                key: “is_new_feature_enabled”
```

Comment:
Could you also give examples of how values are set in providers, e.g. local switches? Right now we only see how `name` and `key` would be defined.

Reply:
value will be retrieved from the provider via commuted property as described above in the example implementation of remote feature switch. For other types of switches it will be the same, the only difference is a different type of provider, which can be probably defined as a generic type parameter instead of injected instance as you suggested above, though I think injecting instance allows better flexibility rather than type that we will need to reconstruct or have a singleton or something else.